### DISCUSSION TITLE
Repository structure

### DISCUSSION THREAD

**Discussion by flanakin** [microsoft/finops-toolkit#862] [General]
Hey, everyone. As part of some discussions around new capabilities we'd like to add to the toolkit, we started discussing the folder hierarchy in the repository. This is more of a conversation for contributors than anything, but I wanted to put it out there to get everyone's thoughts as we're really trying to make this a community-driven project 🙂 Be forewarned that this is a bit technical...

Here's the folder structure we have today:

> 📁 docs
> 📁 docs-wiki
> 📂 src
> ├─ 📁 bicep-registry
> ├─ 📁 open-data
> ├─ 📁 optimization-engine
> ├─ 📁 power-bi
> ├─ 📁 powershell
> ├─ 📂 templates
> │ &nbsp;&nbsp; └─ 📁 finops-hub
> └─ 📂 workbooks
> &nbsp; &nbsp; &nbsp; ├─ 📁 governance
> &nbsp; &nbsp; &nbsp; └─ 📁 optimization
> 📄 package.json

All the docs are in one folder and source code is in another. Each has subfolders per tool. We started with this approach because everything is released as a single package, which has a `package.json` at the root that has the version number for everything.

We're adding a website to manage toolkit solutions, which will have unique dependencies and deployment scripts and resources compared to other solutions. Because of this, we're debating whether the website should be in the common `src` folder or if it should be separate and have its own structure. This then raises a question about the other tools as well. To visualize it, these are the options we discussed:

<table>
<tr><th>One docs + one src</th><th>Group docs/src by tool</th></tr>
<tr><td>
📁 docs<br>
📁 docs-wiki<br>
📂 src<br>
├─ 📁 (tool)<br>
│ &nbsp;&nbsp; └─ 📄 ... (source)<br>
├─ 📂 scripts<br>
│ &nbsp;&nbsp; ├─ 📄 Build-(tool).ps1<br>
│ &nbsp;&nbsp; ├─ 📄 Build-Website.ps1<br>
│ &nbsp;&nbsp; └─ 📄 ...<br>
└─ 📂 website<br>
&nbsp; &nbsp; &nbsp; ├─ 📄 index.htm<br>
&nbsp; &nbsp; &nbsp; └─ 📄 ...<br>
📄 package.json<br>
</td><td>
📁 docs-wiki<br>
📂 (tool)<br>
├─ 📁 docs<br>
├─ 📂 scripts<br>
│ &nbsp;&nbsp; └─ 📄 Build-(tool).ps1<br>
└─ 📁 src<br>
📂 scripts<br>
└─ 📄 ...<br>
📁 website<br>
├─ 📂 scripts<br>
│ &nbsp;&nbsp; └─ 📄 Build-Website.ps1<br>
└─ 📁 src<br>
│ &nbsp;&nbsp; ├─ 📄 index.htm<br>
│ &nbsp;&nbsp; └─ 📄 ...<br>
└─ 📄 package.json (website only)<br>
📄 package.json (whole repo)<br>
</td></tr>
</table>

Hopefully, this visual helps 😕 We're essentially debating whether to group all tools in a single docs and source folder or to separate docs and source by tool.

I'd love to know what others think. I can share another visual if it'd help, but don't want to make it even more confusing 😜 

What does everyone think?

**Comment by helderpinto** [comment]
Maybe not relevant for the discussion, but can you provide more context about what a "website to manage toolkit solutions" is? Is it something we can use to deploy a solution right from the website? Or is it more about managing internal solution stuff, such as versioning, packaging, releasing, etc.?

Regarding the proposed structure, maybe the "One docs + one src" approach is better when you want to deal with all solutions at scale for a specific purpose. For example, I may want to trigger a CI/CD GH workflow only when there is any change in the `src` folder (no matter the solution). The same for the `docs` folder... If we group `docs`/`src` by tool, implementing those GH workflow filters would require manually adding/removing each folder subpath or maybe use a wildcard path (`**`/src/`**`) that might also work, but which probably need more care to maintain...

Otherwise, both approaches are fine as long as the tools hierarchy is as flat as possible. For example, I have some trouble locating things inside the `docs` folder, because of the recent changes (e.g., `_optimize`, `_automation`, `_reporting`, etc. subfolders). A flat hierarchy seems also to be more future-proof (e.g., adding a new tool in the future that does not fit in this hierarchy).

None of these comments are strongly opinionated, though. Looking forward to reading others' comments.

**Reply by flanakin** [reply]
> can you provide more context about what a "website to manage toolkit solutions" is? 

The goal would be something similar to the functionality you would find in the Azure portal. Eventually, we'll seek to add this to the Azure portal, but there are some other things we need to do before we get there. This is a half-step in that direction.

> Regarding the proposed structure, maybe the "One docs + one src" approach is better when you want to deal with all solutions at scale for a specific purpose...

Yep, I had similar thoughts.

Fwiw, there's also a middle option where we separate the website source from the other tool source since there will likely be dependencies there that don't apply to anything else. That said, given the size of the site, we may be overthinking it right now 😜 We can always reorganize content later. This was mostly a thought exercise to get more perspectives before we make a call.

> both approaches are fine as long as the tools hierarchy is as flat as possible.

As an FYI, we are going to flatten it more when we move to MS Learn. I added the grouping to organize the current site more. The website I referenced will also replace the current site we have so it will both introduce tools and allow deploying and managing them. At least, that's the long-term goal.

**Comment by flanakin** [comment]
To close the loop here, we decided to keep the structure we have today (one docs + one src). We can revisit this in the future as needed.
